src/main/scala/com/adobe/api/platform/runtime/mesos/MesosClient.scala

```diff
@@ -692,22 +696,28 @@ object MesosClient {
               .setLaunch(Offer.Operation.Launch.newBuilder
                 .addAllTaskInfos(tasks))))
       .build
-  def updateAgentStats(role: String, offers: Event.Offers) = {
-    val newOfferStats = offers.getOffersList.asScala
-      .filter(_.getResourcesList.asScala.exists(_.getRole == role)) //only include agents with offers that include resources for this role
-      .map(o =>
-        o.getHostname -> AgentStats(
-          o.getResourcesList.asScala
-            .filter(r => r.getRole == role && r.getName == "mem")
-            .foldLeft(0.0)(_ + _.getScalar.getValue),
-          o.getResourcesList.asScala
-            .filter(r => r.getRole == role && r.getName == "cpus")
-            .foldLeft(0.0)(_ + _.getScalar.getValue),
+  def updateAgentStats(config: MesosActorConfig, role: String, offers: Event.Offers) = {
+    //create a map of hostname -> Map[Map[resourceType->resources]], which only includes
+    // - resources with this role
+    // - offers that include all 3 resource types
+    val agentResourceMap = offers.getOffersList.asScala
+      .map(o => o.getHostname -> o.getResourcesList.asScala.filter(_.getRole == role).groupBy(_.getName)) //map hostname to resource map including only resources allocated to this role
+      .filter(a => Set("cpus", "mem", "ports").subsetOf(a._2.keySet)) //remove hosts that do not have resources allocated for all of cpus+mem+ports
+      .toMap
+    //create AgentStats for each agent by summing all
+    val newOfferStats = agentResourceMap.mapValues(
+      v =>
+        AgentStats(
+          v("mem").map(_.getScalar.getValue).sum,
+          v("cpus").map(_.getScalar.getValue).sum,
+          v("ports").foldLeft(0)(_ + _.getRanges.getRangeList.asScala.foldLeft(0)((a, b) =>
+            a + b.getEnd.toInt - b.getBegin.toInt + 1)),
           Instant.now()))
-
+    //update agentOfferHistory (replace existing agents' data, add new agents' data)
     agentOfferHistory = agentOfferHistory ++ newOfferStats
 
-    //prune stats that are > 3min old
-    agentOfferHistory = agentOfferHistory.filter(_._2.lastSeen.isAfter(Instant.now().minusSeconds(180)))
+    //prune stats that are > TTL old (remove old/unseen agents' data)
+    agentOfferHistory =
+      agentOfferHistory.filter(_._2.lastSeen.isAfter(Instant.now().minusSeconds(config.agentStatsTTL.toSeconds)))
```

Comment:
May be pre compute the instant time and use that for filtering

Reply:
Sorry missed this before - this maybe should have its own scheduled task to prune, since this currently assumes offers keep coming, in order for a pruning to happen.
But I'm not sure how you mean to "precompute"? We need to compute the expiration date for each pruning, whether done before or after the pruning is triggered, right?

Reply:
I think latest changes handle this better - scheduled akka task now does pruning, and expiration date is stored along with date of last receipt

Reply:
Is there any risk of conflicting stats - older and newer stats existing at the same time - that could confuse the scheduling in any way ? i.e.a `TTL` of `65 seconds` with a static `10s` prune checking interval.

Reply:
_same time_ updates to stats are handled by actor message sequential processing, and we only update the stats during the message handler, so I don't think this is a risk.